src/main/scala/scorex/core/network/NetworkController.scala

```diff
@@ -224,7 +224,12 @@ class NetworkController(settings: NetworkSettings,
         log.debug(s"Looking for a new random connection")
         val randomPeerF = peerManagerRef ? RandomPeerExcluding(connections.values.flatMap(_.peerInfo).toSeq)
         randomPeerF.mapTo[Option[PeerInfo]].foreach { peerInfoOpt =>
-          peerInfoOpt.foreach(peerInfo => self ! ConnectTo(peerInfo))
+          peerInfoOpt.foreach { peerInfo =>
```

Comment:
1. if the peer fails the further check (I think you're referring to `connectionForPeerAddress(remote).isEmpty && !unconfirmedConnections.contains(remote)`), then it means that it's either become an active connection or it's still in the handshaking process (so still in the `unconfirmedConnections` set) that has a timeout. If we don't put this new check, then we try to connect to an already connected node (so the new duplicated connection is removed) or to a node in the handshaking process (when the first end, the enqueued will fail). In both ways, we would waste a connection attempt (next is scheduled 5 seconds after), and end up removing the peer because we failed to connect to it;
2. if you're referring to `RandomPeerExcluding` and `connectionForPeerAddress`, they do very similar work. Still, I think it was placed there to avoid the scenario when the connection is not excluded by the `RandomPeerExcluding` (e.g. it's not an active connection yet, but its handshaking flow is pending), but when we get to it and try to connect the handshaking process is now completed and it's become an active connection.

Reply:
as was agreed offline:
let's add the unit tests, then see if we can move some of the checks directly to the strategy class